examples/psbt.rs

```diff
@@ -0,0 +1,711 @@
+
+    let got = serialize_hex(&psbt_1);
+    assert_eq!(got, expected_psbt_hex);
+}
+
+/// Signs `psbt` with `keys` if required.
+fn sign(mut psbt: Psbt, keys: &[PrivateKey]) -> Psbt {
+    let secp = SECP256K1;
+
+    let mut inputs = psbt.inputs.clone();
+
+    for (i, input) in inputs.iter_mut().enumerate() {
+        let required_pks: Vec<_> = input.bip32_derivation.keys().collect();
+        for sk in keys {
+            let pk = sk.public_key(&secp);
+            if required_pks.contains(&&pk.inner) {
+                sign_psbt(&sk, &mut psbt, i).expect("failed to sign input with key");
+            }
+        }
+    }
+
+    psbt
+}
+
+/// Signs `psbt` input at `input_index` using `sk`.
+// TODO: Consider adding this code to rust-bitcoin.
+fn sign_psbt(
+    sk: &PrivateKey,
+    psbt: &mut Psbt,
+    input_index: usize,
+) -> Result<(), SignError> {
+    let secp = SECP256K1;
+
+    if input_index >= psbt.inputs.len() || input_index >= psbt.unsigned_tx.input.len() {
+        return Err(SignError::InputIndexOutOfRange);
+    }
+
+    if psbt.inputs[input_index].final_script_sig.is_some()
+        || psbt.inputs[input_index].final_script_witness.is_some()
+    {
+        return Ok(());
+    }
+
+    let pubkey = sk.public_key(secp);
+    if psbt.inputs[input_index].partial_sigs.contains_key(&pubkey) {
+        return Ok(());
+    }
+
+    let (hash, sighash_type) = match psbt.inputs[input_index].witness_utxo {
+        Some(_) => segwit_v0_sighash(psbt, input_index)?,
+        None => legacy_sighash(psbt, input_index)?,
+    };
+
+    // From BIP: Before signing a non-witness input, the Signer must verify that the TXID of the
+    // non-witness UTXO matches the TXID specified in the unsigned transaction.
+    if let Some(tx) = &psbt.inputs[input_index].non_witness_utxo {
+        if tx.txid() != psbt.unsigned_tx.input[input_index].previous_output.txid {
+            return Err(SignError::InvalidTxid);
+        }
+    }
+
+    // From BIP: Before signing a witness input, the Signer must verify that the witnessScript (if
+    // provided) matches the hash specified in the UTXO or the redeemScript, and the redeemScript
+    // (if provided) matches the hash in the UTXO.
+    if let Some(tx) = &psbt.inputs[input_index].witness_utxo {
+        if let Some(script) = &psbt.inputs[input_index].witness_script {
+            let redeem_script = psbt.inputs[1].redeem_script.clone().unwrap();
+
+            let witness_program = Script::new_witness_program(WitnessVersion::V0, &script.wscript_hash());
+            if redeem_script != witness_program {
+                return Err(SignError::WitnessRedeemMismatch);
+            }
+
+
+            let redeem_script = psbt.inputs[1].redeem_script.clone().unwrap();
+            let script = Script::new_p2sh(&redeem_script.script_hash());
+
+            if script != tx.script_pubkey {
+                return Err(SignError::ScritpPubkeyMismatch);
+            }
+
```

Comment:
The two checks above were just hacked together while looking at the test vector PSBTs, this code is not general enough (as I'm finding in #940), before the signing code can be considered ready for `rust-bitcoin` this needs improving and testing with other valid PSBTs, that in itself is an argument for leaving the signing code here in this file for now?

Reply:
I think it makes sense that the generalized code lives in bdk. The question is would using those in examples be a circular dependency. I don't think it makes sense for bdk code to live in integration tests.

Reply:
Oh we definitely cannot depend on bdk. What I mean is that I'm pretty sure this code should live in the `psbt` module but I'm not sure its complete enough yet or if it is, as it currently exists in this PR and #940, general enough. The main stuff I'm concerned about is the BIP174 checks that I added.

Reply:
Why does this signer need to be general if it only serves to test PSBT? Is there an expectation it is library ready as well?

Reply:
I was just wondering that's all, we are about to try to introduce duplicate code in this one and #940 which feels a bit rude. I did #946 to see what others think.